sdk/src/main/java/com/mapbox/maps/viewannotation/ViewAnnotationManagerImpl.kt

```diff
@@ -0,0 +1,253 @@
+      mapboxMap.removeViewAnnotation(id)
+      annotation.view.viewTreeObserver.removeOnGlobalLayoutListener(annotation.globalLayoutListener)
+      mapView.removeView(annotation.view)
+    }
+    currentViewsDrawnMap.clear()
+    positionDescriptorMap.clear()
+    idsToRepositionSet.clear()
+    idsToAddSet.clear()
+    idsToDeleteSet.clear()
+    annotationMap.clear()
+    idLookupMap.clear()
+  }
+
+  private fun validateOptions(options: ViewAnnotationOptions) {
+    checkNotNull(options.geometry) { "Geometry can not be null!" }
+  }
+
+  private fun prepareViewAnnotation(inflatedView: View, options: ViewAnnotationOptions): View {
+    val inflatedViewLayout = inflatedView.layoutParams as FrameLayout.LayoutParams
+    val updatedOptions = options.toBuilder()
+      .width(options.width ?: inflatedViewLayout.width)
+      .height(options.height ?: inflatedViewLayout.height)
+      .build()
+    val viewAnnotation = ViewAnnotation(
+      view = inflatedView,
+      handleVisibilityAutomatically = options.visible == null,
+      viewLayoutParams = inflatedViewLayout,
+    )
+    val globalLayoutListener = ViewTreeObserver.OnGlobalLayoutListener {
+      if (viewAnnotation.handleVisibilityAutomatically) {
+        val isVisibleNow = inflatedView.visibility == View.VISIBLE
+        if (mapboxMap.getViewAnnotationOptions(viewAnnotation.id).value?.visible != isVisibleNow) {
+          mapboxMap.updateViewAnnotation(
+            viewAnnotation.id,
+            ViewAnnotationOptions.Builder()
+              .visible(isVisibleNow)
+              .build()
+          )
+        }
+      }
+    }
+    viewAnnotation.globalLayoutListener = globalLayoutListener
+    inflatedView.viewTreeObserver.addOnGlobalLayoutListener(globalLayoutListener)
+    annotationMap[viewAnnotation.id] = viewAnnotation
+    idLookupMap[inflatedView] = viewAnnotation.id
+    mapboxMap.addViewAnnotation(viewAnnotation.id, updatedOptions)
+    return inflatedView
+  }
+
+  private fun findByFeatureId(featureId: String): Pair<View?, ViewAnnotationOptions?> {
+    annotationMap.forEach { (id, annotation) ->
+      mapboxMap.getViewAnnotationOptions(id).value?.let { options ->
+        if (options.associatedFeatureId == featureId) {
+          return annotation.view to options
+        }
+      }
+    }
+    return null to null
+  }
+
+  private fun drawAnnotationViews(
+    positionDescriptorList: List<ViewAnnotationPositionDescriptor>
+  ) {
+    positionDescriptorMap.clear()
+    idsToRepositionSet.clear()
+    idsToAddSet.clear()
+    idsToDeleteSet.clear()
+
+    positionDescriptorList.forEachIndexed { i, descriptor ->
+      positionDescriptorMap[descriptor.identifier] =
+        ViewPosition(
+          leftTopCoordinate = descriptor.leftTopCoordinate,
+          width = descriptor.width,
+          height = descriptor.height,
+          zIndex = ViewPlugin.VIEW_PLUGIN_Z_TRANSLATION - 1f + i.toFloat() / positionDescriptorList.size
```

Comment:
Can you elaborate on why `zIndex` is calculated in this way?

Reply:
using completely different logic now, PTAL